### PR TITLE
Add host.json for sync trigger payload

### LINKED ISSUE #7288: Adding Durable storageProvider.type to SyncTriggers payload
#### What problem would the feature you're requesting solve? Please describe.
Currently the host does syncTriggers to update the system status of the triggers using this code 
https://github.com/Azure/azure-functions-host/blob/6b95141c6463d8815567bba913346dc780c40630/src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs#L509-L535

But as far as I can tell, the type of the storage provider isn't taken into account. We will need that information in the payload somewhere if the type is `MicrosoftSQL`

#### Describe the solution you'd like
Include the storage type somewhere in the payload. 

#### Describe alternatives you've considered
We could assume `MicrosoftSQL` for envorinments where it should be the right choice, but it may not be entirely correct or lead to issues.  

/cc @mathewc @cgillum

### PR DESCRIPTION
**Pull request by TsuyoshiUshio**
Under the Kubernetes Environment, We need to `host.json` payload for SyncTrigger scenario for transforming Durable Functions and LogicApp scenario. 

I currently testing this branch not finished, so that I make it Draft, However, this change should requires discussion with us. I just simply add the `host.json` payload on the current sync trigger payload. 

CC @pragnagopa , @ahmelsayed 

### Issue describing the changes in this PR

resolves https://github.com/Azure/azure-functions-host/issues/7288 https://github.com/Azure/azure-functions-host/issues/7382

### Pull request checklist

* [x] My changes **do not** require documentation changes
* [ ] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] I have added all required tests (Unit tests, E2E tests)

### CODE REVIEW

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (cgillum) [inline src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Why do you separate `ex.Message` from `ex.StackTrace`? Isn't it better to just trace the full exception details?
```suggestion
                _logger.LogWarning($"Unable to find host configuration file '{scriptPath}'. : {ex}");
```

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (cgillum) [inline src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Just return a new JObject here and remove line 374.
```suggestion
                return new JObject();
```

**Review by cgillum — DISMISSED** [review]
This looks good to me, but let's also get a sign-off from @mathewc

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (TsuyoshiUshio) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: I think we have one more place to need to do new JObject, so that I though to remove the duplication.

Comment (TsuyoshiUshio) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: I update the PR according to your practice. How about this? @cgillum 
It looks CI fails because of the testing for SyncTrigger. I'll check the tests and test with Kuberentes environment and remove draft tomorrow morning.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Note that the extended format { triggers: [], functions: [], secrets: {} } is best effort, and if the payload is oversize based on what FE can handle, we revert to the old triggers array format. This is fine because existing usages of the additional info are able to handle this - the extra data just enables optimizations.

However, in your case, it seems the host config level trigger info you want to add isn't optional? I.e. what what happen in your scenarios if due to size the payload reverted to minimal format?

Comment (mathewc) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Also I worry about dumping the entire host.json file contents. Aren't we really only interested in host level extensions config?

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: See the comment I left in the previous iteration. Looks like a revision you posted blew the comment away.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (TsuyoshiUshio) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Thank you @mathewc for the review. I didn't know the restriction. 
I updated the PR. We deicide to narrow it down the scope. Our usage is, under the Kubernetes Environment, The Sync Trigger will post request to the Kudu Lite API. (As far as I know, It doesn't limit the payload size).  https://github.com/Azure-App-Service/KuduLite/blob/feature/k8se/Kudu.Services/Function/FunctionController.cs#L43 In this case, looks no problem? I'm not sure which part is required for Durable/Logic Apps usage. @cgillum  Do you know which part is required?

Comment (cgillum) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: @mathewc It sounds like you're saying we can't rely on this logic to actually work? What controls whether we use the extended format or the "old triggers array format"? Also, what is the maximum size allowed by the FE? Indeed, the host.json metadata is required for our scenarios. We could probably trim it down to just the ["extensions" section](https://docs.microsoft.com/en-us/azure/azure-functions/functions-host-json#sample-hostjson-file), though, if you're worried about too much host.json bloat.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (divyagandhisethi) [inline src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs:309]: Can we add a unit test specifically for Kubernetes scenario?

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (mathewc) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: The code [here](https://github.com/Azure/azure-functions-host/blob/d831b969dcd3c8ed6a6949b82329b0db7ababf12/src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs#L349) enforces the FE size limit. What we'd have to do is redefine what the "minimal format" is, and rather than just returning the triggers [] in such cases, we'd return `{ triggers: [], extensions: [] }`, scoping it so that we're not dumping the entire host.json in there, but only the extensions config section.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (TsuyoshiUshio) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs:309]: Thank you for you advice! Clever. Done.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (mathewc) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Also note that there are still some legacy codepaths in Kudu (e.g. [here](https://github.com/projectkudu/kudu/blob/0fbbb59fd1151f61daee544715e1326a8632be76/Kudu.Core/Helpers/PostDeploymentHelper.cs#L266) that will end up doing an "old format" sync). @ankitkumarr did the work in Kudu to forward old Kudu synctriggers calls to the Functions runtime to get the new format, but if that fails for any reason, Kudu falls back to it's previous logic where the sync formatting was done by Kudu. So any consumers of this synctriggers payload (ScaleController, Geo ARM APIs, etc.) need to handle both formats, and they do currently.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (cgillum) [inline src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs:309]: Is there a strong reason to limit this to Kubernetes? I'd like to also have this change for Antares too.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: There's already logic in ReadDurableTaskConfig that is reading host.json. Let's not duplicate this code, instead refactoring to share the existing logic.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (mathewc) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Let's not make this Kubernetes specific, since part of our larger goal is to address https://github.com/Azure/azure-functions-host/issues/7288. Instead, I think we should add just the host.json "extensions" section here, and below if we revert to minimal format based on size, we redefine the minimal format as `{ triggers: [], extensions: {} }` ensuring this data is present in that case as well.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (TsuyoshiUshio) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs:309]: Hi @cgillum 
This is mainly time pressure. Since we need to finish today with testing, 
Scoping it that is not effect existing scenario. Also save time for fixing E2E fails. 
We can create a issue to tracking make this global.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (TsuyoshiUshio) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: What do you think, @cgillum ? Is it unblock the Durable for Kubernetes Environment?

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (TsuyoshiUshio) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Thank you  @mathewc , @cgillum @ankitkumarr 
Since I create this PR for unblock the Durable usage,  Is this looks good? or any idea/discussion?

1. limit the payload adding `extensions` part only for kubernetes environmnet.
2. Only the kubernetes environmnet, disable the MaxTriggersStringLength (it is 204800)
3. Create an issue for make SyincTrigger with host.json payload in global for keep on discussing it. 

Only thing, what I don't know is the [code](https://github.com/projectkudu/kudu/blob/0fbbb59fd1151f61daee544715e1326a8632be76/Kudu.Core/Helpers/PostDeploymentHelper.cs#L266) that is done by @ankitkumarr  is releated kubernets environmnet. We are working on this [branch](https://github.com/Azure-App-Service/KuduLite/tree/feature/k8se)

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (pragnagopa) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs:309]: We need to scope the changes for Kubernetes Env as we will be taking commit as hotfix and skip OGFs

@TsuyoshiUshio  - Please open a separate issue to remove the check for Kubernetes env after we cut the hotfix release - to ensure change go through full functions release validation

Comment (cgillum) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs:309]: Makes sense. Thanks!

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (cgillum) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Since this PR is more of an "emergency" PR and we're opening a second one for the "proper" fix, I suggest we defer refactoring for the later PR (let's add a note so that we don't forget).

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Don't need the ? null check, since you're not handling the null if it happened on the next line anyways. Existing ReadDurableTaskConfig code doesn't check for null either.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Recommend using FileUtility.ReadAsync as the other code does, which will enable you to mock the file access as other tests do

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (TsuyoshiUshio) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Thanks.  I change the logic to use the same as Durable.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (TsuyoshiUshio) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Thanks. I didn't know that. I changed it.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (TsuyoshiUshio) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Thank you @mathewc  
Done. 
Could you have a look?

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (TsuyoshiUshio) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: I added some TODO's try not to forget.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Remove this file not found warning completely. Presence of the host.json file isn't mandatory so we don't want to create unnecessary warnings. See the other code which does a FileUtility.FileExists(hostJsonPath) check. 

Also, you should just return null if no host.json file or no extensions node found in the file, and have the calling code only add the extensions property when needed.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Recommend removing this TODO, since likely we'll never want the limit to apply in Kubernetes cases, since there is no FE limit to deal with?

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Please create a new issue in this repo for removing/fixing this code, and link to that issue here. E.g. single comment like "TEMP: refactor this code to properly add extensions in all scenarios (#1234)". We can then include all the details in the issue.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (TsuyoshiUshio) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Done!

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (TsuyoshiUshio) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Thank you for the review. I added issue, and update the comment as you requested! 
https://github.com/Azure/azure-functions-host/issues/7394

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (TsuyoshiUshio) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: I introduce the change as you requested. Thank you for the review!

**Comment by TsuyoshiUshio** [comment]
Hi @mathewc 
I made change as you requested. Thank you for the quick response and great review!

**Comment by TsuyoshiUshio** [comment]
I'm testing this PR E2E once it finished, I'll share it.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Please remove this log statement - it'll just be noise.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (mathewc) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Looks good. I just had one more comment that I added on removing the unnecessary log statement. They we can get this checked in after you add a test for the scenario - I believe you're in the process of adding one?

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (pragnagopa) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: If this needed for debugging - please change to Trace level

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (pragnagopa) [inline src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: take `_applicationHostOptions` as input param to make easier to add unit test

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (mathewc) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: I'd argue that its not needed at all. Note that on the other existing path for Durable where we're reading host.json, we don't log anything. I'd just remove it :)

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (TsuyoshiUshio) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Thank you @mathewc 
I tested E2E, so that code is good for now, I'm planning add unit testing for it. 
Do we need E2E testing as well? If so I'd happy to do that.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (TsuyoshiUshio) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Ok, I removed it.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (pragnagopa) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: For the scope of the changes here unit test is enough! Thanks!

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (TsuyoshiUshio) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Done.

File: src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs
Comment (TsuyoshiUshio) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: I added an Unit Test. Looks good?

Comment (pragnagopa) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: I don’t see updates to unit tests. Missed a commit?

Comment (TsuyoshiUshio) [reply src/WebJobs.Script.WebHost/Management/FunctionsSyncManager.cs]: Hi @pragnagopa 
Sorry, I missed. Now added.

File: test/WebJobs.Script.Tests/Managment/FunctionsSyncManagerTests.cs
Comment (pragnagopa) [inline test/WebJobs.Script.Tests/Managment/FunctionsSyncManagerTests.cs:100]: Add try… finally . Move test directory clean up into finally

Comment (TsuyoshiUshio) [reply test/WebJobs.Script.Tests/Managment/FunctionsSyncManagerTests.cs:100]: That's a good idea.

Comment (TsuyoshiUshio) [reply test/WebJobs.Script.Tests/Managment/FunctionsSyncManagerTests.cs:100]: Done.

**Review by pragnagopa — APPROVED** [review]

File: test/WebJobs.Script.Tests/Managment/FunctionsSyncManagerTests.cs
Comment (mathewc) [inline test/WebJobs.Script.Tests/Managment/FunctionsSyncManagerTests.cs:17]: Why did you add a new file for this test - we already have an existing test suite here https://github.com/Azure/azure-functions-host/blob/6cef5808f90e3a9d3437209ab128361852120597/test/WebJobs.Script.Tests.Integration/Management/FunctionsSyncManagerTests.cs#L31.

In the upcoming PR, let's fix this. Also, the test doesn't really test the feature/changes you added - ensuring that the actual sync triggers payload contains the expected content.

Comment (pragnagopa) [reply test/WebJobs.Script.Tests/Managment/FunctionsSyncManagerTests.cs:17]: Thanks @mathewc  - captured notes here: https://github.com/Azure/azure-functions-host/issues/7394

**Pull request merged**
